src/core/ddsi/src/ddsi_raweth.c

```diff
@@ -168,7 +168,7 @@ static int ddsi_raweth_conn_locator (ddsi_tran_factory_t fact, ddsi_tran_base_t
   {
     loc->kind = NN_LOCATOR_KIND_RAWETH;
     loc->port = uc->m_base.m_base.m_port;
-    memcpy(loc->address, uc->m_base.m_base.gv->extloc.address, sizeof (loc->address));
+    memcpy(loc->address, uc->m_base.m_base.gv->interfaces[0].loc.address, sizeof (loc->address));
```

Comment:
This function is used for setting `pp->m_locator` (q_entity.c:1087), which should get the `extloc` because it is (with specific config settings) used in discovery and for getting the locators that are passed to the receive threads, which should get the internal address (`loc`). Both TCP and UDP implementations are using `loc`, so I agree with changing this also for raw eth. But `pp->m_locator` needs the external address, so maybe add a 'ext' flag to this function?

Reply:
I'm wondering if the "real" problem isn't that the per-participant socket is listening on all interfaces but isn't treated similarly to the other cases in `get_participant_builtin_topic_data`:
```C
    if (pp->e.gv->config.many_sockets_mode == DDSI_MSM_MANY_UNICAST)
    {
      locators_add_one (&def_uni, &pp->m_locator, NN_LOCATOR_PORT_INVALID);
      locators_add_one (&meta_uni, &pp->m_locator, NN_LOCATOR_PORT_INVALID);
    }
    else
    {
      for (int i = 0; i < pp->e.gv->n_interfaces; i++)
      {
        if (!pp->e.gv->xmit_conns[i]->m_factory->m_enable_spdp)
        {
          // skip any interfaces where the address kind doesn't match the selected transport
          // as a reasonablish way of not advertising iceoryx locators here
          continue;
        }
        // FIXME: should have multiple loc_default_uc/loc_meta_uc or compute ports here
        locators_add_one (&def_uni, &pp->e.gv->interfaces[i].extloc, pp->e.gv->loc_default_uc.port);
        locators_add_one (&meta_uni, &pp->e.gv->interfaces[i].extloc, pp->e.gv->loc_meta_uc.port);
      }
    }
```
Should that not be something like:
```C
      for (int i = 0; i < pp->e.gv->n_interfaces; i++)
      {
        if (!pp->e.gv->xmit_conns[i]->m_factory->m_enable_spdp)
          continue;
        if (pp->e.gv->config.many_sockets_mode == DDSI_MSM_MANY_UNICAST)
          defport = metaport = pp->m_locator.port;
        else
        {
          defport = pp->e.gv->loc_default_uc.port;
          metaport = pp->e.gv->loc_meta_uc.port;
        }
        locators_add_one (&def_uni, &pp->e.gv->interfaces[i].extloc, defport);
        locators_add_one (&meta_uni, &pp->e.gv->interfaces[i].extloc, metaport);
      }
```
Then it always advertises the same set of addresses regardless of the `many_sockets_mode`, and it only affects the port numbers. That seems far more sensible than changing the set of addresses as well.

Reply:
Although I don't know all details of the different many-sockets modes, it sounds good to make the set of advertised addresses consistent with non-msm-many mode.